tests/datastructures/test_plan.py

```diff
@@ -0,0 +1,162 @@
+
+import compas
+import pytest
+
+from compas.geometry import Frame
+from compas_fab.datastructures import Action
+from compas_fab.datastructures import IntegerIdGenerator
+from compas_fab.datastructures import Plan
+from compas_fab.datastructures import DependencyIdException
+
+
+def test_action_data():
+    name = "action"
+    parameters = {'param1': 1, 'param2': Frame.worldXY()}
+    action = Action(name, parameters)
+    other_action = Action.from_data(action.data)
+    assert action.name == other_action.name
+    assert action.parameters == other_action.parameters
+
+    action_json = compas.json_dumps(action.data)
+    other_action_data = compas.json_loads(action_json)
+    other_action = Action.from_data(other_action_data)
+    assert other_action.parameters['param2'] == Frame.worldXY()
+
+
+def test_integer_id_generator():
+    generator = IntegerIdGenerator()
+    assert next(generator) == 1
+    assert next(generator) == 2
+    generator_json = compas.json_dumps(generator.data)
+    generator_data = compas.json_loads(generator_json)
+    generator_reincarnate = IntegerIdGenerator.from_data(generator_data)
+    assert next(generator_reincarnate) == 3
+
+
+def test_plan():
+    plan = Plan()
+    assert len(plan.actions) == 0
+
+
+def test_plan_generator_compatibility():
+    action = Action('action', {'param': 3})
+    plan = Plan()
+    action_id = plan.plan_action(action, [])
+    assert action_id == 1
+    assert next(plan._id_generator) == 2
+
+
+def test_plan_with_custom_id_generator():
+    class CustomIdGenerator(object):
+        def __init__(self):
+            self._generator = count(ord('a'))
+
+        def __next__(self):
+            return chr(next(self._generator))
+
+        # alias for ironpython
+        next = __next__
+
+    plan = Plan(CustomIdGenerator())
+    action_a = Action('action_a', {})
+    plan.plan_action(action_a, [])
+    action_b = Action('action_b', {})
+    plan.plan_action(action_b, ['a'])
+    assert 'a' in plan.actions
+    assert 'b' in plan.actions
+    assert ['a'] == plan.get_dependency_ids('b')
+    plan.remove_action('a')
+    assert 'a' not in plan.actions
+    assert 'b' in plan.actions
+    assert [] == plan.get_dependency_ids('b')
+    action_c = Action('action_c', {})
+    plan.append_action(action_c)
+    assert 'c' in plan.actions
+    assert ['b'] == plan.get_dependency_ids('c')
+
+
+def test_plan_data():
+    plan = Plan()
+    plan.append_action(Action('action_1', {'param': Frame.worldXY()}))
+    other_plan = Plan.from_data(plan.data)
+    assert plan.actions.keys() == other_plan.actions.keys()
+
+    # the data attributes point to the same generator,
+    # so no testing `append_action` yet
+    plan_json = compas.json_dumps(plan.data)
+    other_plan_data = compas.json_loads(plan_json)
+    other_plan = Plan.from_data(other_plan_data)
+    assert plan.actions.keys() == other_plan.actions.keys()
+    assert other_plan.get_action(1).parameters['param'] == Frame.worldXY()
```

Comment:
@yck011522 Not sure what you mean about `Action.from_data` not working... This test passes.  What sort of test do you think would fail?